pkg/storage/backup_compaction_iterator.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package storage
+
+import (
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/util"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/errors"
+)
+
+// BackupCompactionIterator wraps a SimpleMVCCIterator and only surfaces the
+// latest valid key of a given MVCC key that is also below the asOfTimestamp, if
+// set. The iterator does surface point tombstones, but not any MVCC keys
+// shadowed by tombstones below the asOfTimestamp, if set. The iterator assumes
+// that it will not encounter any write intents.
+type BackupCompactionIterator struct {
+	iter SimpleMVCCIterator
+
+	// asOf is the latest timestamp of a key surfaced by the iterator.
+	asOf hlc.Timestamp
+
+	// valid tracks if the current key is valid
+	valid bool
+
+	// err tracks if iterating to the current key returned an error
+	err error
+}
+
+var _ SimpleMVCCIterator = &BackupCompactionIterator{}
+
+func NewBackupCompactionIterator(
+	iter SimpleMVCCIterator, asOf hlc.Timestamp,
+) *BackupCompactionIterator {
+	if asOf.IsEmpty() {
+		asOf = hlc.MaxTimestamp
+	}
+	return &BackupCompactionIterator{
+		iter: iter,
+		asOf: asOf,
+	}
+}
+
+func (f *BackupCompactionIterator) Close() {
+	f.iter.Close()
+}
+
+func (f *BackupCompactionIterator) Next() {
+	f.NextKey()
+}
+
+func (f *BackupCompactionIterator) NextKey() {
+	f.iter.NextKey()
+	f.advance()
+}
+
+func (f *BackupCompactionIterator) SeekGE(originalKey MVCCKey) {
+	synthetic := MVCCKey{Key: originalKey.Key, Timestamp: f.asOf}
+	f.iter.SeekGE(synthetic)
+	if f.advance(); f.valid && f.UnsafeKey().Less(originalKey) {
+		f.NextKey()
```

Comment:
leave a comment along the lines of: "see readAsOfIterator comment for explanation of this"